external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -3626,36 +3690,64 @@ fn method_call(
 fn method_call_resolve(
     context: &mut Context,
     loc: Loc,
-    edotted: ExpDotted,
+    edotted: &mut ExpDotted,
     method: Name,
     ty_args_opt: Option<Vec<Type>>,
-) -> Option<(ModuleIdent, FunctionName, ResolvedFunctionType, T::Exp)> {
-    use TypeName_ as TN;
-    use Type_ as Ty;
-
-    let mut edotted = edotted;
-
+) -> ResolvedMethodCall {
     edotted.loc = loc;
     let edotted_ty = core::unfold_type(&context.subst, edotted.last_type());
 
-    let tn = match &edotted_ty.value {
-        Ty::Apply(_, tn @ sp!(_, TN::ModuleType(_, _) | TN::Builtin(_)), _) => tn,
+    let Some(tn) = type_to_type_name(context, &edotted_ty, loc, "method call".to_string()) else {
+        return ResolvedMethodCall::InvalidBaseType;
+    };
+    let Some((m, f, fty)) =
+        core::make_method_call_type(context, loc, &edotted_ty, &tn, method, ty_args_opt)
+    else {
+        return ResolvedMethodCall::UnknownName;
+    };
+    let usage = match &fty.params[0].1.value {
+        Type_::Ref(true, _) => DottedUsage::Borrow(true),
+        Type_::Ref(false, _) => DottedUsage::Borrow(false),
+        _ => DottedUsage::Use,
+    };
+    ResolvedMethodCall::Resolved(Box::new(m), f, fty, usage)
+}
+
+fn type_to_type_name(
+    context: &mut Context,
+    ty: &Type,
+    loc: Loc,
+    error_msg: String,
+) -> Option<TypeName> {
+    use TypeName_ as TN;
+    use Type_ as Ty;
+    fn make_ascii_titlecase(in_s: &str) -> String {
```

Comment:
Might be worth throwing this into shared. I think I had some spots around where I did things like `foo(..., "case", "Case")`